### PR TITLE
Allow staff to edit final determination except status

### LINKED ISSUE #635: Ability for team to edit/delete a determination msg with an option to not re-send the applicant notifications
This exists in the previous system. Use case, we need to go back and edit the final determination msg outcome due to error or some other reason but don't want the applicant to get another duplicate notification with the only thing that's changed is the status.

### PR DESCRIPTION
**Pull request by vaibhavmule**
Fixes #635

(This PR now includes commits from branch [enhancement/add_sent_notice_option_determinations](https://github.com/OpenTechFund/hypha/tree/enhancement/add_sent_notice_option_determinations))

I have gone through CreateOrUpdateView for Determination and checked if there is a possibility to use the existing `Update` view for Determination. 

The existing Update view is quite complex, 
- it has created in the mind that when you are updating the determination when in draft state
- has permission-based on if the user can create or edit determination

After going through this, decided to create a new update view for editing Determination.

**Following things need to be done in case you want to update status for Submission.**

There is no provision to revert back to `Ready for determination` stage.

I have looked at existing use cases found that there is a [restart_stage method on Submission](https://github.com/OpenTechFund/hypha/blob/ceb74015590ae24d1412478505faeda6951c080d/hypha/apply/funds/models/submissions.py#L462). 

After that, I have look at [FSMField Docs](https://github.com/viewflow/django-fsm#usage) and [blog](https://hashedin.com/blog/a-guide-to-managing-finite-state-machine-using-django-fsm/),  came with the following function to programmatically change the status of the submission. 

https://github.com/OpenTechFund/hypha/pull/1901/files#diff-dc3c11f62e7001fabe8c26c7ace3a61fR477
```python
@transition(
    status, source='*',
    target=RETURN_VALUE(
        'determination',
        'ext_determination',
        'proposal_determination',
        'concept_determination',
        'com_determination',
    ),
    permission=make_permission_check({UserPermissions.STAFF}),)
def revert_to_determination(self, **kwargs):
    """
    Revert back to determination
    """
    if self.stage == RequestExt:
        return 'ext_determination'
    elif self.stage == RequestCom:
        return 'com_determination'
    elif self.stage == Concept:
        return 'concept_determination'
    elif self.stage == Proposal:
        return 'proposal_determination'

    return 'determination'
```

So before updating the Determination above `revert_to_determination` is called and the following side effects are taken into consideration.
- if the determination was approved, then you change it to dismissed, `you may need to delete the project object created for submission`.
- if the determination was dismissed, then you change it to approved, `you must create a project`.

**For delete determination, we need to implement the above status change part so when we delete the determination, we can revert back status of a submission to` ready for determination.`**

### CODE REVIEW

File: hypha/apply/determinations/views.py [hunk 1]
```diff
@@ -429,3 +431,82 @@ class DeterminationDetailView(ViewDispatcher):
     partner_view = PartnerDeterminationDetailView
     community_view = CommunityDeterminationDetailView
     applicant_view = ApplicantDeterminationDetailView
+
+
+@method_decorator(staff_required, name='dispatch')
+class DeterminationUpdateView(UpdateView):
```
Comment (frjo) [inline hypha/apply/determinations/views.py:437]: Call it "DeterminationEditView" instead to sync up with the naming in reviews.

File: hypha/apply/funds/models/submissions.py [hunk 2]
```diff
@@ -474,6 +478,31 @@ def restart_stage(self, **kwargs):
             return 'invited_to_proposal'
         return INITIAL_STATE
 
+    @transition(
```
Comment (frjo) [inline hypha/apply/funds/models/submissions.py:481]: The state of the submission should not change when the determination is edited. My understanding is that the edits are only for archival purposes.

File: hypha/apply/determinations/forms.py [hunk 3]
```diff
@@ -403,3 +404,18 @@ def __init__(self, *args, submissions, initial={}, **kwargs):
...
 
         self.fields = self.apply_form_settings('proposal', self.fields)
+
+
+class UpdateDeterminationForm(forms.ModelForm):
+    message = RichTextField(label='Determination message')
+
+    class Meta:
+        model = Determination
+        fields = ['outcome', 'message']
```
Comment (frjo) [inline hypha/apply/determinations/forms.py:414]: Remove outcome and add send_notice.

File: hypha/apply/determinations/forms.py [hunk 4]
```diff
@@ -403,3 +404,18 @@ def __init__(self, *args, submissions, initial={}, **kwargs):
         self.fields['outcome'].widget = forms.HiddenInput()
 
         self.fields = self.apply_form_settings('proposal', self.fields)
+
+
+class UpdateDeterminationForm(forms.ModelForm):
```
Comment (frjo) [inline hypha/apply/determinations/forms.py:409]: Rename to "DeterminationEditForm".

File: hypha/apply/determinations/templates/determinations/determination_detail.html [hunk 5]
```diff
@@ -6,7 +6,8 @@
 {% block content %}
 <div class="admin-bar">
     <div class="admin-bar__inner">
-        <h2 class="heading heading--no-margin">Determination {% if determination.is_draft %}[DRAFT]{% endif %}</h2>
+        <h2 class="heading heading--no-margin">Determination {% if determination.is_draft %}[DRAFT]{% else %}<a class="button" href="{% url 'apply:submissions:determinations:update' submission_pk=determination.submission.id pk=determination.id %}">Edit</a>{% endif %}
```
Comment (frjo) [inline hypha/apply/determinations/templates/determinations/determination_detail.html:9]: See how the edit button look and is placed on reviews and copy that.

**Comment by frjo** [comment]
@vaibhavmule I rebased this branch of enhancement/add_sent_notice_option_determinations that in turn was rebased of latest master.

Please to a hard reset to get your local branch updated. I think this should work.

~~~~
git fetch origin
git checkout feature/edit-delete-determination
git reset origin/feature/edit-delete-determination --hard
~~~~

Rebasing instead of merging is a bit of a hassel but you get a nicer commit history that is easier to read and understand.

File: hypha/apply/determinations/forms.py (hunk 3 above)
Comment (frjo) [reply hypha/apply/determinations/forms.py:414]: They also want all the other text fields.

Comment (vaibhavmule) [reply hypha/apply/determinations/forms.py:414]: oh okay, Need to add more changes in that case, as form fields are different for 4 workflows.

File: hypha/apply/determinations/templates/determinations/determination_form.html [hunk 6]
```diff
@@ -3,7 +3,7 @@
 {% block header %}
 <div class="admin-bar">
     <div class="admin-bar__inner">
-        <h2 class="heading heading--no-margin">{% if object %}Update Determination draft{% else %}Add Determination{% endif %}</h2>
+        <h2 class="heading heading--no-margin">{% if object %}Update Determination {% if object.is_draft %}draft{% endif %}{% else %}Add Determination{% endif %}</h2>
```
[the remark this answers is not in the document]
Comment (vaibhavmule) [reply hypha/apply/determinations/templates/determinations/determination_form.html:6]: @frjo Yes, it is the same as the review.
The only difference is here it is `Add Determination` as compare for Review it is `Create Review`
Let me know what do you think of it.

[the remark this answers is not in the document]
Comment (vaibhavmule) [reply hypha/apply/determinations/templates/determinations/determination_form.html:6]: Done.

File: hypha/apply/determinations/forms.py (hunk 3 above)
Comment (vaibhavmule) [reply hypha/apply/determinations/forms.py]: I have used existing form with few modifications.

**Comment by vaibhavmule** [comment]
@frjo I have made the requested changes, Please review and let me know.

**Pull request merged**